File: macos/Sources/ReleaseAssets.swift
import Foundation

struct ReleaseAsset: Equatable {
    let name: String
    let url: URL
    let size: Int64
}

enum ReleaseAssets {
    static let maxBytes: Int64 = 200 * 1024 * 1024

    /// In-app updates always replace `/Applications/MonitorSuhu.app`.
    /// Repo `build/` copies and mounted DMGs must not be the install target —
    /// that left GitHub 1.0.5 "installed" only in the source tree.
    static func installDestination(runningPath: String) -> String {
        let path = URL(fileURLWithPath: runningPath).standardizedFileURL.path
        if path == "/Applications/MonitorSuhu.app" || path.hasPrefix("/Applications/MonitorSuhu.app/") {
            return "/Applications/MonitorSuhu.app"
        }
        if path.hasPrefix("/Applications/"), path.hasSuffix("/MonitorSuhu.app") {
            return path
        }
        return "/Applications/MonitorSuhu.app"
    }

    static func pick(_ assets: [ReleaseAsset], platform: String = "macos") -> ReleaseAsset? {
        assets.first { matches($0.name, platform: platform) && isTrusted($0.url, name: $0.name, platform: platform) }
    }

    static func matches(_ name: String, platform: String) -> Bool {
        let n = name.lowercased()
        guard n.contains("monitorsuhu") else { return false }
        switch platform {
        case "macos": return n.contains("macos") && n.hasSuffix(".dmg")
        case "windows": return n.contains("windows") && n.hasSuffix(".exe")
        case "linux": return n.contains("linux") && (n.hasSuffix(".tar.gz") || n.hasSuffix(".zip"))
        default: return false
        }
    }

    static func isTrusted(_ url: URL, name: String, platform: String) -> Bool {
        guard matches(name, platform: platform) else { return false }
        if name.contains("/") || name.contains("\\") || name.contains("..") { return false }
        guard url.scheme?.lowercased() == "https" else { return false }
        switch url.host?.lowercased() {
        case "github.com", "objects.githubusercontent.com", "release-assets.githubusercontent.com":
            return true
        default:
            return false
        }
    }
}

File: windows/src/MonitorSuhu.Core/ReleaseAssets.cs
namespace MonitorSuhu.Core;

public sealed record ReleaseAsset(string Name, string Url, long Size);

public sealed record GitHubRelease(
    string Tag,
    string? HtmlUrl,
    IReadOnlyList<ReleaseAsset> Assets,
    bool Draft = false,
    bool Prerelease = false);

public static class ReleaseAssets
{
    public const long MaxBytes = 200L * 1024 * 1024;

    public static string CurrentPlatform =>
        OperatingSystem.IsWindows() ? "windows"
        : OperatingSystem.IsMacOS() ? "macos"
        : OperatingSystem.IsLinux() ? "linux"
        : "unknown";

    public static ReleaseAsset? Pick(IEnumerable<ReleaseAsset> assets, string? platform = null)
    {
        var os = (platform ?? CurrentPlatform).Trim().ToLowerInvariant();
        foreach (var asset in assets)
        {
            if (Matches(asset.Name, os) && IsTrusted(asset.Url, asset.Name, os))
                return asset;
        }
        return null;
    }

    public static bool Matches(string name, string platform)
    {
        if (string.IsNullOrWhiteSpace(name)) return false;
        var n = name.ToLowerInvariant();
        if (!n.Contains("monitorsuhu")) return false;
        return platform switch
        {
            "macos" => n.Contains("macos") && n.EndsWith(".dmg"),
            "windows" => n.Contains("windows") && n.EndsWith(".exe"),
            "linux" => n.Contains("linux") && (n.EndsWith(".tar.gz") || n.EndsWith(".zip")),
            _ => false
        };
    }

    public static bool IsTrusted(string url, string name, string platform)
    {
        if (!Matches(name, platform)) return false;
        if (name.IndexOfAny(['/', '\\']) >= 0) return false;
        if (name.Contains("..", StringComparison.Ordinal)) return false;
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri)) return false;
        if (uri.Scheme != Uri.UriSchemeHttps) return false;
        var host = uri.Host.ToLowerInvariant();
        return host is "github.com"
            or "objects.githubusercontent.com"
            or "release-assets.githubusercontent.com";
    }

    /// Newest non-draft release that ships a file for this OS.
    /// A Windows-only tag is not "latest" on macOS or Linux.
    public static (string Tag, string? HtmlUrl, ReleaseAsset Asset)? LatestFor(
        IEnumerable<GitHubRelease> releases,
        string? platform = null)
    {
        var os = (platform ?? CurrentPlatform).Trim().ToLowerInvariant();
        foreach (var release in releases)
        {
            if (release.Draft || release.Prerelease) continue;
            var asset = Pick(release.Assets, os);
            if (asset is null) continue;
            var tag = release.Tag.StartsWith('v') ? release.Tag[1..] : release.Tag;
            return (Versioning.Trim(tag), release.HtmlUrl, asset);
        }
        return null;
    }
}
